crates/wasi-nn/src/tf.rs

```diff
@@ -0,0 +1,406 @@
+
+        // Now, figure out what TF operation to bind this tensor to.
+        let operation = self
+            .graph
+            .operation_by_name_required(&tensor_info.name().name)?;
+
+        // Convert the dimensions to `u64`s.
+        let dims = tensor
+            .dimensions
+            .as_slice()?
+            .iter()
+            .map(|d| *d as u64)
+            .collect::<Vec<_>>();
+
+        // Copy the tensor bytes to the Tensorflow container. We pretend the
+        // tensor has byte elements (though it may contain elements of any
+        // `TensorType`) because we expect the user to provide the tensor in the
+        // exact, compatible byte format for Tensorflow. Ideally we would avoid
+        // the copy here and just point to the original bytes (TODO: investigate
+        // unsafely using `as_mut_ptr`).
+        self.tensors.push(match tensor.type_ {
+            TensorType::F32 => TensorTypes::TTF32(TFTensor::<f32>::new(&dims)),
+            TensorType::F16 => TensorTypes::TTF16(TFTensor::<f32>::new(&dims)),
+            TensorType::U8 => TensorTypes::TTU8(TFTensor::<u8>::new(&dims)),
+            TensorType::I32 => TensorTypes::TTI32(TFTensor::<i32>::new(&dims)),
+        });
+
+        let data = tensor.data.as_slice()?;
+
+        // Assign the tensor to the session arguments. The `add_feed`
+        // documentation says that because most operations have only one output
+        // (and presumably one input), so the input index is likely 0. Note that
+        // we need to do some awkward hoop-jumping here:
+        // - in order to maintain the lifetime of `SessionRunArgs`, which
+        //   borrows the tensor data, we copy the tensor data into our `Self`
+        //   above (otherwise we cannot guarantee that the borrowed data will be
+        //   there when we actually need it, in `compute`).
+        // - but we also must fit within the Wiggle-generated
+        //   `BackendExecutionContext` trait, which says this function must take
+        //   `&mut self`. So we pretend that `&mut self` lives as long as, well,
+        //   itself (`'a`) using `transmute` and use our new `self_` to borrow
+        //   the tensor data we copied to `Self`.
+        let self_ = unsafe { std::mem::transmute::<&mut Self, &'a mut Self>(self) };
+        let tensor_ref = self_.tensors.last_mut().unwrap();
+
+        use TensorTypes::*;
+        match tensor_ref {
+            TTU8(t) => {
+                copy_data(t, &data);
+                self_.args.add_feed(&operation, 0, t);
+            }
+            TTF16(t) => {
+                copy_data(t, &data);
+                self_.args.add_feed(&operation, 0, t);
+            }
+            TTF32(t) => {
+                copy_data(t, &data);
+                self_.args.add_feed(&operation, 0, t);
+            }
+            TTI32(t) => {
+                copy_data(t, &data);
+                self_.args.add_feed(&operation, 0, t);
+            }
+        };
```

Comment:
It's unclear to me why [`copy_from_slice`](https://github.com/abrown/wasmtime/commit/a9ddd19876a366c1509b80a623056bc0e2c0d952#diff-a940d24b7d1898c64717b16afcb125c155b4cb11526ef7e546734d2b1f17b5c0R179) has been replaced by all of these match cases?